File: src/Physics/simplePhysics.py
from PYB11Generator import *
from physics import *

@PYB11template("dim")
class SimplePhysics(Physics):
    def pyinit(self,
               nodeList="NodeList*",
               constants="PhysicalConstants&"):
        return
    # @PYB11virtual
    # def UpdateDerivatives(self,dt="const double"):
    #     return "void"

SimplePhysics1d = PYB11TemplateClass(SimplePhysics,
                              template_parameters = ("1"),
                              cppname = "SimplePhysics<1>",
                              pyname = "SimplePhysics1d",
                              docext = " (1D).")
SimplePhysics2d = PYB11TemplateClass(SimplePhysics,
                              template_parameters = ("2"),
                              cppname = "SimplePhysics<2>",
                              pyname = "SimplePhysics2d",
                              docext = " (2D).")
SimplePhysics3d = PYB11TemplateClass(SimplePhysics,
                              template_parameters = ("3"),
                              cppname = "SimplePhysics<3>",
                              pyname = "SimplePhysics3d",
                              docext = " (3D).") 

File: src/Physics/simplePhysics.cc
#include "physics.hh"
#include <iostream>

template <int dim>
class SimplePhysics : public Physics<dim> {
protected:

public:
    using Vector = Lin::Vector<dim>;
    using VectorField = Field<Vector>;
    using ScalarField = Field<double>;

    SimplePhysics(NodeList* nodeList,
                        PhysicalConstants& constants) :
        Physics<dim>(nodeList,constants) {

        int numNodes = nodeList->size();
        this->template EnrollFields<double>({"y"});
        this->template EnrollStateFields<double>({"y"});
    }

    ~SimplePhysics() {}

    virtual void
    PreStepInitialize() override {
        State<dim> state = this->state;
        NodeList* nodeList = this->nodeList;

        ScalarField* sy       = state.template getField<double>("y");
        ScalarField* y        = nodeList->template getField<double>("y");

        sy->copyValues(y);
    }

    virtual void
    EvaluateDerivatives(const State<dim>* initialState, State<dim>& deriv, const double time, const double dt) override {
        // extremely simple ode, y(t) = 10*t^2 - 1/2*t^3  -> y'=20*t - 3/2*t^2

        NodeList* nodeList = this->nodeList;
        PhysicalConstants constants = this->constants;
        int numNodes = nodeList->size();

        ScalarField* y        = initialState->template getField<double>("y");
        ScalarField* dydt     = deriv.template getField<double>("y");

        #pragma omp parallel for
        for (int i=0; i<numNodes ; ++i) {
            dydt->setValue(i,20.0*(time+dt) - 1.5*(time+dt)*(time+dt));
        }
    }

    virtual void
    FinalizeStep(const State<dim>* finalState) override {
        State<dim> state = this->state;
        NodeList* nodeList = this->nodeList;

        ScalarField* sy       = state.template getField<double>("y");
        ScalarField* fy       = finalState->template getField<double>("y");
        ScalarField* y        = nodeList->template getField<double>("y");

        sy->copyValues(fy);
        y->copyValues(fy);
    }
};
